### PR TITLE
playwright-test 1.47.0 -> 1.47.2 plus update script

### PR DESCRIPTION
**Pull request by humemm**
## Description of changes
This updates ```playwright-test``` to address some regressions that were introduced in 1.47.0 and supplies an update script for automating future updates.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by winterqt — CHANGES_REQUESTED** [review]
At a glance, `nix-update-script` should work fine here; could you try that? Furthermore, even if it doesn't, please split the package update and the addition of the update script into two separate commits.

**Comment by kalekseev** [comment]
We have [update.sh](https://github.com/NixOS/nixpkgs/blob/f7bfe794a79c69c01f2dce20fb7b287ae98a96eb/pkgs/development/python-modules/playwright/update.sh) script for playwright-python that also updates playwright driver to the version used in python module https://github.com/microsoft/playwright-python/blob/65a86d90b6a1643b19544cfe1e2108863a84aa24/setup.py#L33

Can you update existing update.sh script to make it update playwright to latest patch version so instead of using driver version 1.7.0 it will update playwright to 1.7.2?

**Comment by kalekseev** [comment]
@humemm 

> From what I can tell the package you're referring to playwright-python which in nix is called just playwright has a dependency on playwright-test which in nix it's all bundled under playwright-driver along with core and browsers.

See a [discussion](https://github.com/NixOS/nixpkgs/pull/302759#discussion_r1584700163_) regarding wrong names for packages, would be great to address this before next release

>  the python package should not have squatted the top-level `playwright` attribute so I am onboard with the change but that needs mentioning in the release notes.
> It's too late for breaking changes for 24.05 nixos/doc/manual/release-notes/rl-2405.section.md, and nixos/doc/manual/release-notes/rl-2411.section.md does not exist yet so looks like it will have to wait.
> 


>  I think a proper way to do this would be to have a separate playwright-driver under the pythonPackages namespace so that you can track whatever version is appropriate for playwright-python and we can freely update this to the latest as we wish.

In this case nixpkgs will have two sets of playwright and related browsers.
Until we have real evidence that patch version disperancy breaks python playwright I'd keep a single playwright.
Thus in this pr I'd prefer to see update to python module update.sh script that relaxes patch version constraint allowing playwright update from 1.74.0 to 1.74.2.

**Comment by teto** [comment]
> See a https://github.com/NixOS/nixpkgs/pull/302759#discussion_r1584700163_ regarding wrong names for packages, would be great to address this before next release


https://github.com/NixOS/nixpkgs/pull/302759#issuecomment-2044795462 might be a more precise link

Fully agree. I would prefer a brutal "playwright" -> playwright-test renaming with a nentry in 24.11 release notes but the sensible way might be instead to, over one release, have playwright still pointing at plawyright-test with a warning "in 25.04, this attribute will refer to 'playwright' and not the python package anymore". And in 25.05 we do the switch. I am not sure "playwright" has many consumers so it could be fine . This kind of question is why we have a release coordinator I guess :D @RossComputerGuy

**Comment by kalekseev** [comment]
@humemm 

> Would it be better if we had separate update scripts for each package?

No. As I said we should have the same minor version in playwright and playwright python but we can move update.sh from python module into playwright module.

**Comment by kalekseev** [comment]
For reference here's a list of PRs https://github.com/microsoft/playwright-python/pulls?q=roll where playwright version updated in python module. Minor version updates such as 1.46 -> 1.47 usually have quite significant changes, while patch versions usually updated without any changes, thus we should stick with the same minor version in both packages.

**Comment by kalekseev** [comment]
@humemm playwright and playwright-python developed by microsoft, minor version updates for them usually released within days, so there's no reason to have two sets of playwright drivers and browsers.

> As I pointed out playwright introduced some pretty significant regressions in 1.47.0 that were addressed in 1.47.2. 

I already said that we can relax dependency of python module and upgrade playwright to latest patch version, which is 1.47.2.

**Comment by kalekseev** [comment]
> Or now that I think about it's probably better to keep the update steps in the python package as well just in case and only change the version pinning that was in place. Let me know I can change it back if you think it's better that way.

@humemm I think we should do something like this:

1. (optional) Move update.sh into playwright module
2. Modify update.sh to get driver_version from playwright-python but bump it to the latest patch version
3. Use latest patch version to upgrade playwright

The rest of the script should be the same, it should update both packages but use latest patch version for playwright.

**Comment by kalekseev** [comment]
@humemm the change doesn't handle a situation when playwright has new minor version in latest release, eg. python module uses driver version 1.47.0 and latest playwright is version 1.48.0, in this case the script must update playwright to 1.47.2 not 1.48.0

> The updated script takes in some positional arguments that are not documented at all and passes an invalid flag to nix store namely --unpack that throws errors and results in hashes for browsers to be stripped and set to an empty string.

The upack was added recently, you probably use older version of nix https://nix.dev/manual/nix/2.24/command-ref/new-cli/nix3-store-prefetch-file#opt-unpack

**Comment by kalekseev** [comment]
Actually nixpkgs stuck on nix 2.18 because of regressions https://github.com/NixOS/nixpkgs/issues/343147 so it might be that update script wont work on nixpkgs-update CI, I don't know which version of nix it uses. Would be nice to replace `nix store --upack` with something that works on v2.18.

**Comment by kalekseev** [comment]
> > @humemm the change doesn't handle a situation when playwright has new minor version in latest release, eg. python module uses driver version 1.47.0 and latest playwright is version 1.48.0, in this case the script must update playwright to 1.47.2 not 1.48.0
> 
> @kalekseev on this point I'm not sure if you saw my comment above. I tested this scenario by moving playwright-python back to 1.46.0 and keeping playwright on 1.47.2 and it worked just fine. I know you mentioned that there were a lot of new features introduced in 1.47.0 for python but the API is backwards compatible for minor versions. I just don't see a reason to complicate the logic in the update script and potentially introduce bugs for a problem that doesn't exist. As you mentioned in your comment earlier in addition to that Microsoft coordinates those larger releases so I don't think we risk running into a problem.

@humemm check this prs https://github.com/microsoft/playwright-python/pulls?q=roll they are prs with playwright version update in python-playwright, if you check minor versions updates you can easily see non trivial api changes. Considering that python package usually released within several days of playwright minor release we should stay on the same minor version in nixpkgs.

**Comment by kalekseev** [comment]
nixpkgs is not npm, 10 days delay is quite good actually.

File: pkgs/development/python-modules/playwright/update.sh [hunk 1]
```diff
@@ -4,16 +4,29 @@ set -euo pipefail
 
 root="$(dirname "$(readlink -f "$0")")"
 
-version=$(curl ${GITHUB_TOKEN:+" -u \":$GITHUB_TOKEN\""} -s https://api.github.com/repos/microsoft/playwright-python/releases/latest | jq -r '.tag_name | sub("^v"; "")')
-# Most of the time, this should be the latest stable release of the Node-based
-# Playwright version, but that isn't a guarantee, so this needs to be specified
-# as well:
-setup_py_url="https://github.com/microsoft/playwright-python/raw/v${version}/setup.py"
-driver_version=$(curl -Ls "$setup_py_url" | grep '^driver_version =' | grep -Eo '[0-9]+\.[0-9]+\.[0-9]+')
```
Comment (kalekseev) [inline pkgs/development/python-modules/playwright/update.sh:12]: you should take driver version from setup.py

https://github.com/microsoft/playwright-python/blob/d9cdfbb1e178b6770625e9f857139aff77516af0/setup.py#L33

File: pkgs/development/python-modules/playwright/update.sh [hunk 2]
```diff
@@ -4,16 +4,29 @@ set -euo pipefail
...
+# Get all versions of Node-based Playwright
+all_driver_versions=$(curl ${GITHUB_TOKEN:+" -u \":$GITHUB_TOKEN\""} -s https://api.github.com/repos/microsoft/playwright/releases | jq -r '.[].tag_name | sub("^v"; "")')
+
+# Find the latest driver version that matches the major.minor of playwright-python
+driver_version=$(echo "$all_driver_versions" | grep "^${python_major_minor}" | sort -V | tail -n 1)
+
+# Fallback to the latest version if no match is found (optional)
+if [[ -z "$driver_version" ]]; then
+  driver_version=$(curl ${GITHUB_TOKEN:+" -u \":$GITHUB_TOKEN\""} -s https://api.github.com/repos/microsoft/playwright/releases/latest | jq -r '.tag_name | sub("^v"; "")')
+fi
```
Comment (kalekseev) [inline pkgs/development/python-modules/playwright/update.sh:22]: fallback to driver version in setup.py

**Review by kalekseev — CHANGES_REQUESTED** [review]

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Please cleanup the git history. Also please drop the merge commits.

Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions) when naming your commits.

**Comment by teto** [comment]
this has been bugging me for a while so I opened https://github.com/NixOS/nixpkgs/pull/346269 to prepare the "playwright" attribute change

**Pull request closed without merging**